internal/dockerhost/dockerhost.go

```diff
@@ -0,0 +1,187 @@
+	if err := d.Client.List(ctx, dockerHosts, listOption); err != nil {
+		// return the default dockerhost
+		return dockerHost, fmt.Errorf("unable to list dockerhosts in the lagoon namespace, there may be none or something went wrong: %v", err)
+	}
+
+	// check available dockerhosts
+	availableHosts := []string{}
+	sort.Slice(dockerHosts.Items, func(i, j int) bool { return dockerHosts.Items[i].Name < dockerHosts.Items[j].Name })
+	for _, dh := range dockerHosts.Items {
+		dhsvc := fmt.Sprintf("%s.%s.svc", dh.Name, d.DockerHostNamespace)
+		for _, port := range dh.Spec.Ports {
+			// network policy may need to be adjusted in chart to support a controller installed in a non lagoon namespace
+			if tcpCheck(dhsvc, strconv.Itoa(int(port.Port))) {
+				availableHosts = append(availableHosts, dhsvc)
+			}
+		}
+	}
+
+	if cacheHost, ok := d.ReuseCache.Get(reuseIdentifier); ok {
+		if helpers.ContainsString(availableHosts, cacheHost) {
+			// if the namespace has already got a host assigned, prefer it
+			dockerHost = cacheHost
+		}
+	}
+
+	// pick a host to use
+	hostsInUse := countDupes(d.BuildCache.Values())
+	dockerHost = d.pickHost(dockerHost, availableHosts, hostsInUse, qosMax, qos)
+	// finally assign in the cache
+	_ = d.BuildCache.Add(buildName, dockerHost)
+	_ = d.ReuseCache.Add(reuseIdentifier, dockerHost)
+	return dockerHost, nil
+}
+
+func New(client client.Client,
+	dockerHostNamespace,
+	reuseType string,
+	reuseCache,
+	buildCache *lru.Cache[string, string],
+) *DockerHost {
+	return &DockerHost{
+		Client:              client,
+		DockerHostNamespace: dockerHostNamespace,
+		ReuseType:           reuseType,
+		ReuseCache:          reuseCache,
+		BuildCache:          buildCache,
+	}
+}
+
+func tcpCheck(host string, port string) bool {
+	timeout := 5 * time.Second
+	conn, err := net.DialTimeout("tcp", net.JoinHostPort(host, port), timeout)
+	if err != nil {
+		fmt.Println(err)
+		return false
+	}
+	if conn != nil {
+		defer conn.Close()
+	}
+	return true
+}
+
+func countDupes(list []string) map[string]int {
+	dupeFreq := make(map[string]int)
+	for _, item := range list {
+		// check if the item/element exist in the map
+		_, exist := dupeFreq[item]
+		if exist {
+			dupeFreq[item] += 1 // increase counter by 1 if already in the map
+		} else {
+			dupeFreq[item] = 1 // else start counting from 1
+		}
+	}
+	return dupeFreq
+}
+
+// pick host will attempt to determine the most suitable docker host to use with limited information
+// without deep inspecting the docker hosts for actual usage/consumption, just use the number of builds as an estimate
+func (d *DockerHost) pickHost(chosenHost string, availableHosts []string, hostsInUse map[string]int, qosMax int, qos bool) string {
+	buildsPerHost := 0
+	numAvailableHosts := len(availableHosts)
+	if numAvailableHosts == 0 {
+		// return the default docker-host if no available hosts
+		return fmt.Sprintf("docker-host.%s.svc", d.DockerHostNamespace) // default dockerhost
```

Comment:
you've done this default docker host logic in more than one place - it might make sense to just wrap it in a function so that it becomes
`return d.defaultDockerHost()`